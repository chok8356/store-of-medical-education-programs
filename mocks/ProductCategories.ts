import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
