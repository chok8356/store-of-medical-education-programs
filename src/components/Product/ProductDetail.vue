<template>
  <b-card
    v-if="product"
    class="overflow-hidden">
    <b-row>
      <b-col md="6">
        <b-card-img :src="product.img" />
      </b-col>
      <b-col md="6">
        <b-card-body>
          <h1 class="mb-4">
            {{ product.name }}
          </h1>

          <b-card-text>
            {{ product.description }}
          </b-card-text>

          <b-badge variant="info">
            {{ category.name }}
          </b-badge>

          <div class="d-block mt-4">
            <b-button
              variant="primary"
              @click="addToCart">
              В корзину
            </b-button>
          </div>
        </b-card-body>
      </b-col>
    </b-row>
  </b-card>
</template>

<script lang="ts">
import { BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductDetail',
  components: {
    BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg
  }
})
export default class ProductDetail extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct
  @Prop({ type: Object as PropType<ICategory> }) category!: ICategory

  addToCart() {
    this.$emit('add-to-cart')
  }
}
</script>
