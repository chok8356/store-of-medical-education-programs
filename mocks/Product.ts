import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
