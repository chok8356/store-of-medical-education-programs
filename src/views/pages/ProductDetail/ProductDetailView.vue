<template>
  <div class="product-detail-view">
    <b-button
      class="mb-4"
      @click="backToStore">
      Назад в магазин
    </b-button>

    <div
      v-loading="isProductLoading"
      class="product-detail-view__loader" />

    <ProductDetail
      v-if="detailProduct && category"
      :category="category"
      :product="detailProduct"
      @add-to-cart="onAddToCart" />
  </div>
</template>

<script lang="ts">
import { BButton } from 'bootstrap-vue'
import { Vue, Component } from 'vue-property-decorator'
import { namespace } from 'vuex-class'
import ProductDetail from '@/components/Product/ProductDetail.vue'
import { loading } from '@/directives'
import { ICategory } from '../../../../types/ICategory'
import { IProduct } from '../../../../types/IProduct'

const ProductStore = namespace('product')
const CartStore = namespace('cart')
const CategoryStore = namespace('category')

@Component({
  name: 'ProductDetailView',
  components: { ProductDetail, BButton },
  directives: {
    loading
  }
})
export default class ProductDetailView extends Vue {
  @ProductStore.State detailProduct: IProduct | null
  @ProductStore.Action getDetailProduct: any

  @CategoryStore.State categories: ICategory[]
  @CategoryStore.Action getCategories: any

  @CartStore.Action addToCart: any

  isProductLoading = true

  async created() {
    if (!this.categories.length) await this.getCategories()
    if (this.productIdFromRoute) await this.getDetailProduct(this.productIdFromRoute)
    this.isProductLoading = false
  }

  get category() {
    return this.categories.find(x => x.id === this.detailProduct?.category)
  }

  get productIdFromRoute() {
    const id = this.$route?.params?.id
    return id ? parseInt(id) : null
  }

  backToStore() {
    this.$router.push({ name: 'ProductsList' })
  }

  onAddToCart() {
    if (this.detailProduct) this.addToCart(this.detailProduct.id)
  }
}
</script>
