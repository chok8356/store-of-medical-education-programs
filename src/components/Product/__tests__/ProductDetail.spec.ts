import { shallowMount } from '@vue/test-utils'

import { products } from '@/../mocks/Product'
import { categories } from '@/../mocks/ProductCategories'
import ProductDetail from '../ProductDetail.vue'

const product = products[0]
const category = categories[1]

describe('ProductDetail', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(ProductDetail, {
      propsData: {
        product,
        category
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('exist props', () => {
    expect(wrapper.props('product')).toStrictEqual(product)
    expect(wrapper.props('category')).toStrictEqual(category)
  })

  it('emit add to cart', () => {
    wrapper.vm.addToCart()
    expect(wrapper.emitted('add-to-cart')).toBeDefined()
  })
})
