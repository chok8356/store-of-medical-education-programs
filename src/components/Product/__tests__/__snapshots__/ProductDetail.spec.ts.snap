// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail render correctly 1`] = `
<b-card-stub
  bodytag="div"
  class="overflow-hidden"
  footertag="div"
  headertag="div"
  subtitletag="h6"
  subtitletextvariant="muted"
  tag="div"
  titletag="h4"
>
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      md="6"
      tag="div"
    >
      <b-card-img-stub
        src="https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds="
      />
    </b-col-stub>
     
    <b-col-stub
      md="6"
      tag="div"
    >
      <b-card-body-stub
        bodytag="div"
        subtitletag="h6"
        subtitletextvariant="muted"
        titletag="h4"
      >
        <h1
          class="mb-4"
        >
          
          Продукт 1
        
        </h1>
         
        <b-card-text-stub
          texttag="p"
        >
          
          1 Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.
        
        </b-card-text-stub>
         
        <b-badge-stub
          tag="span"
          target="_self"
          variant="info"
        >
          
          Категория 2
        
        </b-badge-stub>
         
        <div
          class="d-block mt-4"
        >
          <b-button-stub
            tag="button"
            target="_self"
            type="button"
            variant="primary"
          >
            
            В корзину
          
          </b-button-stub>
        </div>
      </b-card-body-stub>
    </b-col-stub>
  </b-row-stub>
</b-card-stub>
`;
